src/parser/jobs/rdm/Procs.js

```diff
@@ -0,0 +1,169 @@
+	}
+	//Global timestamp tracking per Proc
+	_currentProcs = {}
+	_castState = null
+	_impactfulProcOverride = false
+	_previousCast = null
+	_bossWasInvuln = false
+
+	constructor(...args) {
+		super(...args)
+
+		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('applybuff', {
+			to: 'player',
+			abilityId: PROCS,
+		}, this._onGain)
+		this.addHook('removebuff', {
+			to: 'player',
+			abilityId: PROCS,
+		}, this._onRemove)
+		this.addHook('refreshbuff', {
+			to: 'player',
+			abilityId: PROCS,
+		}, this._onRefresh)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		const abilityID = event.ability.guid
+		const invuln = this.downtime.getDowntime(this._previousCast||0, event.timestamp)
+		this._previousCast = event.timestamp
+		this._bossWasInvuln = invuln > 0
+
+		if (abilityID === ACTIONS.IMPACT.id && this._currentProcs[STATUSES.IMPACTFUL.id]) {
+			const impactRemainingDuration = event.timestamp - (this._currentProcs[STATUSES.IMPACTFUL.id] || 0)
+			if (impactRemainingDuration <= IMPACT_OVERRIDE_THRESHOLD) {
+				//Use Impactful at 8 or less seconds remaining, regardless of all other procs
+				this._impactfulProcOverride = true
+			}
+		}
+
+		this._castState = abilityID
+	}
+
+	_onGain(event) {
+		const statusID = event.ability.guid
+
+		//Initialize if not present
+		if (!(statusID in this._history)) {
+			this._history[statusID] = {
+				overWritten: 0,
+				invuln: 0,
+				missed: 0,
+				wasted: 0,
+			}
+		}
+
+		if (!(statusID in this._currentProcs)) {
+			this._currentProcs[statusID] = 0
+		}
+
+		this._currentProcs[statusID] = event.timestamp
+	}
+
+	_onRefresh(event) {
+		const statusID = event.ability.guid
+
+		if (this._currentProcs[statusID] > 0 && !this._impactfulProcOverride && statusID !== STATUSES.ENHANCED_SCATTER.id) {
+			this._history[statusID].overWritten++
+		}
+
+		this._currentProcs[statusID] = event.timestamp
+	}
+
+	_onRemove(event) {
+		const statusID = event.ability.guid
+
+		if (this._bossWasInvuln) {
+			this._history[statusID].invuln++
+		} else if (this._castState !== this._castStateMap[statusID].id) {
+			this._history[statusID].missed++
+		}
+
+		if (this._castState === this._castStateMap[statusID].id) {
+		//Reset this statusID!
+			this._currentProcs[statusID] = 0
+		}
+	}
+
+	_onComplete() {
+		for (const statusID in this._history) {
+			//Leaving here for future debug purposes.
+			//const util = require('util')
+			// console.log(statusID)
+			// console.log(STATUSES[statusID].name)
+			// console.log(util.inspect(this._history[statusID], {showHidden: true, depth: null}))
+			this.generateSuggestions(statusID)
+		}
+	}
+
+	generateSuggestions(statusID) {
+		if (this._history[statusID].missed) {
+			this.suggestions.add(new TieredSuggestion({
+				icon: this._castStateMap[statusID].icon,
+				why: `${this._history[statusID].missed} ${this._castStateMap[statusID].name}${this._history[statusID].missed !== 1 ? 's' : ''} casts missed due to buff falling off`,
```

Comment:
For future PRs - might be worth  pulling a lot of these lookups out into a few consts at the start of the function. Will shorten up the code a lot, and make the i18n strings look nicer when that gets implemented.